activitysim/benchmarking/componentwise.py

```diff
@@ -0,0 +1,601 @@
+            "multiprocess component benchmarking is not yet implemented"
+        )
+    else:
+        open_pipeline(resume_after, mode="r")
+
+    for k in preload_injectables:
+        if inject.get_injectable(k, None) is not None:
+            logger.info("pre-loaded %s", k)
+
+    # Directories Logging
+    for k in ["configs_dir", "settings_file_name", "data_dir", "output_dir"]:
+        logger.info(f"DIRECTORY {k}: {inject.get_injectable(k, None)}")
+
+    # Settings Logging
+    log_settings = [
+        "checkpoints",
+        "chunk_training_mode",
+        "chunk_size",
+        "chunk_method",
+        "trace_hh_id",
+        "households_sample_size",
+        "check_for_variability",
+        "use_shadow_pricing",
+        "want_dest_choice_sample_tables",
+        "log_alt_losers",
+    ]
+    for k in log_settings:
+        logger.info(f"SETTING {k}: {config.setting(k)}")
+
+    logger.info("setup_component completed: %s", component_name)
+
+
+def run_component(component_name):
+    logger.info("run_component: %s", component_name)
+    try:
+        if config.setting("multiprocess", False):
+            raise NotImplementedError(
+                "multiprocess component benchmarking is not yet implemented"
+            )
+            # logger.info('run multiprocess simulation')
+            #
+            # from activitysim.core import mp_tasks
+            # injectables = {k: inject.get_injectable(k) for k in INJECTABLES}
+            # mp_tasks.run_multiprocess(injectables)
+            #
+            # assert not pipeline.is_open()
+            #
+            # if config.setting('cleanup_pipeline_after_run', False):
+            #     pipeline.cleanup_pipeline()
+        else:
+            run_model(component_name)
+    except Exception as err:
+        logger.exception("run_component exception: %s", component_name)
+        raise
+    else:
+        logger.info("run_component completed: %s", component_name)
+    return 0
+
+
+def teardown_component(component_name):
+    logger.info("teardown_component: %s", component_name)
+
+    # use the pipeline module to clear out all the orca tables, so
+    # the next benchmark run has a clean slate.
+    # anything needed should be reloaded from the pipeline checkpoint file
+    pipeline_tables = pipeline.registered_tables()
+    for table_name in pipeline_tables:
+        logger.info("dropping table %s", table_name)
+        pipeline.drop_table(table_name)
+
+    if config.setting("multiprocess", False):
+        raise NotImplementedError("multiprocess benchmarking is not yet implemented")
+    else:
+        pipeline.close_pipeline()
+    logger.critical(
+        "teardown_component completed: %s\n\n%s\n\n", component_name, "~" * 88
+    )
+    return 0
+
+
+def pre_run(
+    model_working_dir,
+    configs_dirs=None,
+    data_dir="data",
+    output_dir="output",
+    settings_file_name=None,
+):
+    """
+    Pre-run the models, checkpointing everything.
```

Comment:
What does pre-running the models mean and why is it needed?  More documentation here.